### ISSUE TITLE
Time eviction issue

### ISSUE THREAD

**Issue by laurentvaills** [ben-manes/caffeine#159]
Hi, 

I tried to use Caffeine and the time based eviction. 
I started with a simple example, to try it but I can't make it work as I expect. Maybe you can help. 
Here is my Java class that uses Caffeine : 
```java
public class TestCaffeine2 {

    static interface CacheableObject {

        /**
         * Time-to-live of this object, in nanoseconds
         *
         * @return Time-to-live of this object, in nanoseconds
         */
        long getTtl();
    }

    static class MyCacheableString implements CacheableObject {

        private String value;
        private long ttl;

        MyCacheableString(String value, long ttl) {
            this.value = value;
            this.ttl = ttl;
        }

        String getValue() {
            return value;
        }

        public long getTtl() {
            return ttl;
        }
    }

    static class Repository {

        private ScheduledExecutorService scheduledExecutorService;
        private final Random random;

        Repository(ScheduledExecutorService scheduledExecutorService, Random random) {
            this.scheduledExecutorService = scheduledExecutorService;
            this.random = random;
        }

        CompletableFuture<MyCacheableString> get(String key) {
            System.out.println("Looking up for " + key);

            CompletableFuture<MyCacheableString> completableFuture = new CompletableFuture<>();
            // This is a simulation of an asynchronous call to fetch the object
            // matching the given key from an external data storage
            scheduledExecutorService.schedule((Runnable) () -> completableFuture.complete(new MyCacheableString(key + "@" + System
                                                      .currentTimeMillis(), TimeUnit.NANOSECONDS.convert(1, TimeUnit.SECONDS))),
                                              random.nextInt(5),
                                              TimeUnit.SECONDS);

            return completableFuture;
        }
    }

    void run(String[] args) throws InterruptedException {
        ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(2);
        Random random = new Random(System.currentTimeMillis());
        Repository repository = new Repository(scheduledExecutorService, random);

        final AsyncLoadingCache<String, MyCacheableString> cache;
        cache = Caffeine.newBuilder()
                        .expireAfter(new Expiry<String, MyCacheableString>() {
                            public long expireAfterCreate(String key,
                                                          MyCacheableString myCacheableString,
                                                          long currentTime) {
                                System.out.println("ExpireAfterCreate : " + currentTime);
                                return currentTime + myCacheableString.getTtl();
                            }

                            public long expireAfterUpdate(String key, MyCacheableString myCacheableString,
                                                          long currentTime, long currentDuration) {
                                System.out.println("ExpireAfterUpdate : " + currentTime + " " + currentDuration);
                                return currentDuration;
                            }

                            public long expireAfterRead(String key, MyCacheableString myCacheableString,
                                                        long currentTime, long currentDuration) {
                                System.out.println("ExpireAfterRead   : " + currentTime + " " + currentDuration);
                                return currentDuration;
                            }
                        })
                        .buildAsync((key, executor) -> repository.get(key));

        CompletableFuture<MyCacheableString> object1 = cache.get("foo");
        System.out.println("Got the 1st object " + object1);
        object1.thenAccept((myCacheableString) -> System.out.println("Got " + myCacheableString.getValue()));

        TimeUnit.SECONDS.sleep(12);

        // Now, the cached object should have been evicted
        CompletableFuture<MyCacheableString> object2 = cache.get("foo");
        System.out.println("Got the 2nd object " + object2);
        object2.thenAccept((myCacheableString) -> System.out.println("Got " + myCacheableString.getValue()));

        if (object1 == object2) {
            System.out.println("!!! The object was not evicted !!!");
        }

        scheduledExecutorService.shutdown();
        System.out.println("END");
    }

    public static void main(String[] args) throws InterruptedException {
        new TestCaffeine2().run(args);
    }
}
```
My object is never evicted from the cache. I am also quite surprised to never see my trace for "ExpireAfterCreate".  
What did I do wrong ?

**Comment by ben-manes** [comment]
`expireAfterCreate` should return a duration and not a point in time. In your code change it to return `myCacheableString.getTtl()`. Also note that for async caches, the timestamps are relative to when the future completed rather than when inserted.

It could be simpler to use a `FakeTicker` and `Awaitility` to model the state changes in the test and avoid any races.

Async is confusing so there may be a bug causing it to call notify on update instead of create. I have to debug further to understand the flow. This is a new feature so there may be bugs that my tests didn't catch.

**Comment by laurentvaills** [comment]
I tried both the point in time and a duration before submitting the issue but none worked. 
Indeed, I also think that there may be a bug that calls notify on update instead of create.

**Comment by ben-manes** [comment]
Can you try with a synchronous cache? I probably won't have time to look into the async problem for a few days, but will try to have a fix out by next Monday.

**Comment by laurentvaills** [comment]
I confirm it works with a synchronous cache. (And I see the trace for `ExpireAfterCreate`).

**Comment by ben-manes** [comment]
In the async case your expiry is wrapped by `AsyncExpiry` to translate the call. When the future is in-flight the duration is infinity (Long.MAX_VALUE). When it completes an callback updates it, which should calls `expireAfterUpdate`. The wrapper should notice that the creation was never called due to being close to infinity (larger than MAX_VALUE / 2) and delegate to `expireAfterCreate` instead. If a subsequent update occurs, e.g. `cache.put`, then it should delegate to an update call. Clearly that is not working as intended.

**Comment by ben-manes** [comment]
If I trace with a debugger it works. So perhaps there is a race in reading a field value.

**Comment by ben-manes** [comment]
I think that I've found the bug, which is due to an incorrect infinite time set on [addition](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1336). The logic here is so that an in-flight future does not expire prematurely, though that block may no longer be necessary. It should have the infinite timestamp set already, but in prior releases was a little more racy (expiration was set outside of an entry's synchronized block).

Unfortunately that was accidentally set to the duration `MAXIMUM_EXPIRY` (150 years), which is the highest valid duration that is honored (and half of Long.MAX_VALUE). Anything higher is used to signal that the future was in-flight and to adapt the calls correctly (e.g. update -> create). Setting it to `Long.MAX_VALUE` passes your test, as does removing the block.

I reused the `expireAfterWrite` tests which sets a new duration on create or update. If I had tested on create only then I think this would fail without any threading needed. I'll try to get coverage for that configuration.

I'll have to think through if the problematic block is even necessary. I'll have a release for you by the weekend, if that's okay?

**Comment by laurentvaills** [comment]
Thanks for the update and for your working chasing that bug. 
I can wait until next week for a new release or pull-request to try.

**Comment by ben-manes** [comment]
Fixed. I'll release tonight.

The patch passes your test and I added expire-after-create to the suite.

**Comment by ben-manes** [comment]
Released 2.5.2

**Comment by laurentvaills** [comment]
I confirm that is working for me. 
Thanks for your reactivity.
